td-shim-tools/src/bin/td-payload-reference-calculator/main.rs

```diff
@@ -31,6 +43,92 @@ fn param(param: &str, size: &str) -> Result<String> {
     padding_digest(param, siz)
 }
 
+fn qemu(path: &str, size: &str) -> Result<String> {
+    let path = Path::new(path).to_path_buf();
+    let siz = parse::<u64>(size)?;
+    let file_size = std::fs::metadata(&path)?.len();
+    if file_size > siz {
+        bail!("File size should be less than `kernel-size`");
+    }
+    let buf = std::fs::read(path)?;
+    qemu_patch(buf)
+}
+
+fn qemu_patch(mut buf: Vec<u8>) -> Result<String> {
+    // patching type_of_loader @0x210
+    buf[0x210] = 0xb0;
+
+    // patching loadflags @0x211
+    buf[0x211] = 0x81;
+
+    //patching heap_end_ptr @0x224 cmdline_addr - real_addr - 0x200 = 0xfe00
+    buf[0x224] = 0x00;
+    buf[0x225] = 0xfe;
+
+    //patching cmd_line_ptr @0x228 cmdline_addr = 0x20000
+    buf[0x228] = 0x00;
+    buf[0x229] = 0x00;
+    buf[0x22A] = 0x02;
+
+    let mut hasher = sha2::Sha384::new();
+    let image = buf.clone();
+    let (number_of_region_entry, regions_base, regions_size) = get_image_regions(image);
+
+    for index in 0..number_of_region_entry {
+        hasher.update(&buf[regions_base[index]..regions_base[index] + regions_size[index]]);
+    }
+
+    let res = hasher.finalize();
+    Ok(hex::encode(res))
+}
+
+fn get_image_regions(buf: Vec<u8>) -> (usize, Vec<usize>, Vec<usize>) {
```

Comment:
```suggestion
fn get_image_regions(buf: &[u8]) -> (usize, Vec<usize>, Vec<usize>) {
```

Use `&[u8]` than `Vec<u8>` could help to avoid extra memory copy.

Reply:
Thank you for your suggestions! 
I've added the link and replaced `Vec<u8>` with `&[u8]`.

Reply:
I cloned the code from https://github.com/tianocore/edk2-staging/tree/2024-tdvf-ww01.4/OvmfPkg. To build TDVF there are two ways [in the document](https://github.com/tianocore/edk2-staging/blob/2024-tdvf-ww01.4/OvmfPkg/IntelTdx/README#L20), which one do I need to use?

Also, could you help to give some guides to build the qemu of the intended version? I'd like to do some test.

Reply:
> To build TDVF there are two ways [in the document](https://github.com/tianocore/edk2-staging/blob/2024-tdvf-ww01.4/OvmfPkg/IntelTdx/README#L20), which one do I need to use?

Config-B is better, it will extend eventlog to the RTMR or TPM.

>  could you help to give some guides to build the qemu of the intended version?

I think you can follow the qemu wiki, the latest qemu version should support TDVF. My version is v7.2.0 

> Is the kernel vmlinuz (compressed) specified by qemu, and the TDVF will uncompress and then boot?

For this question, to be honest, I'm not an expert :(  
I think the kernel `vmlinuz` may be uncompressed by qemu, TDVF  just load and boot. I can do more research and investigation to figure it out :) .

Reply:
Hi, I use qemu from https://github.com/qemu/qemu.git with tag `v7.2.0` and tdvf suggested by you. When I run
```
qemu-system-x86_64 -machine confidential-guest-support=tdx0 \
    -object tdx-guest,id=tdx0 \
    -drive if=pflash,format=raw,unit=0,file=./edk2-staging/Build/IntelTdx/DEBUG_GCC5/FV/OVMF_CODE.fd \
    -drive if=pflash,format=raw,unit=1,file=./edk2-staging/Build/IntelTdx/DEBUG_GCC5/FV/OVMF_VARS.fd
```

Error raised 
```
qemu-system-x86_64: -object tdx-guest,id=tdx0: Parameter 'qom-type' does not accept value 'tdx-guest'
```
What should I  do?

Reply:
Hi @Xynnn007 , I got the message that the qemu in my test env was supplied from internel linux os team and BKC team, and the qemu from https://github.com/qemu/qemu.git  don't have tdx-guest support.

Link for qemu-tdx: https://github.com/intel/qemu-tdx/tree/tdx-qemu-upstream
Link for bkc: https://github.com/intel/qemu-td-partitioning (Note: the qemu-bkc is private repo)

You can get support from above link.

Reply:
Hi @OuyangHang33 . I tested locally with the [qemu](https://github.com/intel/qemu-tdx/tree/tdx-qemu-upstream) and [tdvf](https://github.com/tianocore/edk2-staging/tree/2024-tdvf-ww01.4/OvmfPkg) with the following command
```shell
./qemu-system-x86_64 \
    -machine confidential-guest-support=tdx0 \
    -object tdx-guest,id=tdx0,sept-ve-disable=off \
    -drive if=pflash,format=raw,unit=0,file=.../edk2-staging/Build/IntelTdx/DEBUG_GCC5/FV/OVMF_CODE.fd \
    -drive if=pflash,format=raw,unit=1,file=.../edk2-staging/Build/IntelTdx/DEBUG_GCC5/FV/OVMF_VARS.fd
```

and the result is
```
qemu-system-x86_64: ../system/vl.c:2684: qemu_machine_creation_done: Assertion `machine->cgs->ready' failed.
Aborted (core dumped)
```

Is there anything wrong?


Also, if I use CoCo/Kata's tdx-qemu to run this firmware, I met the error on containerd side
```
qemu-system-x86_64-tdx-experimental: TDVF metadata doesn't specify TD_HOB location.
```

Reply:
Did you test with the TDX supported system? Because I got a training this afternoon, I will try your command and tdx-qemu on my side later and give the result before tomorrow evening.

Reply:
Yes. It is on tdx machine.

thanks! take your time.